File: cAER/tags/caer-0.9.5/ext/sshs/sshs_helper.c
#include "sshs_internal.h"

static void sshsHelperAllocSprintf(char **strp, const char *format, ...) __attribute__ ((format (printf, 2, 3)));

// Put NULL in *strp on failure (memory allocation failure).
static void sshsHelperAllocSprintf(char **strp, const char *format, ...) {
	va_list argptr;

	va_start(argptr, format);
	size_t printLength = (size_t) vsnprintf(NULL, 0, format, argptr);
	va_end(argptr);

	*strp = malloc(printLength + 1);
	if (*strp == NULL) {
		return;
	}

	va_start(argptr, format);
	vsnprintf(*strp, printLength + 1, format, argptr);
	va_end(argptr);
}

// Return NULL on unknown type. Do not free returned strings!
const char *sshsHelperTypeToStringConverter(enum sshs_node_attr_value_type type) {
	// Convert the value and its type into a string for XML output.
	switch (type) {
		case BOOL:
			return ("bool");

		case BYTE:
			return ("byte");

		case SHORT:
			return ("short");

		case INT:
			return ("int");

		case LONG:
			return ("long");

		case FLOAT:
			return ("float");

		case DOUBLE:
			return ("double");

		case STRING:
			return ("string");

		default:
			return (NULL); // UNKNOWN TYPE.
	}
}

// Return -1 on unknown type.
enum sshs_node_attr_value_type sshsHelperStringToTypeConverter(const char *typeString) {
	if (typeString == NULL) {
		return (-1); // NULL STRING.
	}

	// Convert the value string back into the internal type representation.
	if (strcmp(typeString, "bool") == 0) {
		return (BOOL);
	}
	else if (strcmp(typeString, "byte") == 0) {
		return (BYTE);
	}
	else if (strcmp(typeString, "short") == 0) {
		return (SHORT);
	}
	else if (strcmp(typeString, "int") == 0) {
		return (INT);
	}
	else if (strcmp(typeString, "long") == 0) {
		return (LONG);
	}
	else if (strcmp(typeString, "float") == 0) {
		return (FLOAT);
	}
	else if (strcmp(typeString, "double") == 0) {
		return (DOUBLE);
	}
	else if (strcmp(typeString, "string") == 0) {
		return (STRING);
	}

	return (-1); // UNKNOWN TYPE.
}

// Return NULL on failure (either memory allocation or unknown type / faulty conversion).
// Strings returned by this function need to be free()'d after use!
char *sshsHelperValueToStringConverter(enum sshs_node_attr_value_type type, union sshs_node_attr_value value) {
	// Convert the value and its type into a string for XML output.
	char *valueString;

	switch (type) {
		case BOOL:
			// Manually generate true or false.
			if (value.boolean) {
				valueString = strdup("true");
			}
			else {
				valueString = strdup("false");
			}

			break;

		case BYTE:
			sshsHelperAllocSprintf(&valueString, "%" PRIu8, value.ubyte);
			break;

		case SHORT:
			sshsHelperAllocSprintf(&valueString, "%" PRIu16, value.ushort);
			break;

		case INT:
			sshsHelperAllocSprintf(&valueString, "%" PRIu32, value.uint);
			break;

		case LONG:
			sshsHelperAllocSprintf(&valueString, "%" PRIu64, value.ulong);
			break;

		case FLOAT:
			sshsHelperAllocSprintf(&valueString, "%g", value.ffloat);
			break;

		case DOUBLE:
			sshsHelperAllocSprintf(&valueString, "%lg", value.ddouble);
			break;

		case STRING:
			valueString = strdup(value.string);
			break;

		default:
			valueString = NULL; // UNKNOWN TYPE.
			break;
	}

	return (valueString);
}

// Return false on failure (unknown type / faulty conversion), the content of
// value is undefined. For the STRING type, the returned value.string is a copy
// of the input string. Remember to free() it after use!
bool sshsHelperStringToValueConverter(enum sshs_node_attr_value_type type, const char *valueString,
	union sshs_node_attr_value *value) {
	if (valueString == NULL || value == NULL) {
		return (false); // NULL STRING.
	}

	switch (type) {
		case BOOL:
			// Boolean uses custom true/false strings.
			if (strcmp(valueString, "true") == 0) {
				value->boolean = true;
			}
			else {
				value->boolean = false;
			}

			break;

		case BYTE:
			if (sscanf(valueString, "%" SCNu8, &value->ubyte) != 1) {
				return (false); // CONVERSION FAILURE.
			}

			break;

		case SHORT:
			if (sscanf(valueString, "%" SCNu16, &value->ushort) != 1) {
				return (false); // CONVERSION FAILURE.
			}

			break;

		case INT:
			if (sscanf(valueString, "%" SCNu32, &value->uint) != 1) {
				return (false); // CONVERSION FAILURE.
			}

			break;

		case LONG:
			if (sscanf(valueString, "%" SCNu64, &value->ulong) != 1) {
				return (false); // CONVERSION FAILURE.
			}

			break;

		case FLOAT:
			if (sscanf(valueString, "%g", &value->ffloat) != 1) {
				return (false); // CONVERSION FAILURE.
			}

			break;

		case DOUBLE:
			if (sscanf(valueString, "%lg", &value->ddouble) != 1) {
				return (false); // CONVERSION FAILURE.
			}

			break;

		case STRING:
			value->string = strdup(valueString);
			if (value->string == NULL) {
				return (false); // MALLOC FAILURE.
			}

			break;

		default:
			return (false); // UNKNOWN TYPE.
	}

	return (true);
}

File: branches/jAER_2.0/src/main/java/net/sf/jaer2/util/SSHSHelper.java
package net.sf.jaer2.util;

public class SSHSHelper {
	// Return NULL on unknown type. Do not free returned strings!
	public static String typeToStringConverter(final Class<?> type) {
		// Convert the value and its type into a string for XML output.
		if (type == Boolean.class) {
			return ("bool");
		}
		else if (type == Byte.class) {
			return ("byte");
		}
		else if (type == Short.class) {
			return ("short");
		}
		else if (type == Integer.class) {
			return ("int");
		}
		else if (type == Long.class) {
			return ("long");
		}
		else if (type == Float.class) {
			return ("float");
		}
		else if (type == Double.class) {
			return ("double");
		}
		else if (type == String.class) {
			return ("string");
		}
		else {
			return null; // UNKNOWN TYPE.
		}
	}

	// Return -1 on unknown type.
	public static Class<?> stringToTypeConverter(final String typeString) {
		if (typeString == null) {
			return null; // NULL STRING.
		}

		// Convert the value string back into the internal type representation.
		switch (typeString) {
			case "bool":
				return Boolean.class;
			case "byte":
				return Byte.class;
			case "short":
				return Short.class;
			case "int":
				return Integer.class;
			case "long":
				return Long.class;
			case "float":
				return Float.class;
			case "double":
				return Double.class;
			case "string":
				return String.class;
			default:
				return null; // UNKNOWN TYPE.
		}
	}

	// Return NULL on failure (either memory allocation or unknown type / faulty
	// conversion).
	// Strings returned by this function need to be free()'d after use!
	public static String valueToStringConverter(final Class<?> type, final Object value) {
		// Convert the value and its type into a string for XML output.
		if (type == Boolean.class) {
			// Manually generate true or false.
			if ((Boolean) value) {
				return "true";
			}

			return "false";
		}
		else if ((type == Byte.class) || (type == Short.class) || (type == Integer.class) || (type == Long.class)) {
			return value.toString();
		}
		else if ((type == Float.class) || (type == Double.class)) {
			return String.format("%g", value);
		}
		else if (type == String.class) {
			return (String) value;
		}
		else {
			return null; // UNKNOWN TYPE.
		}
	}

	// Return false on failure (unknown type / faulty conversion), the content
	// of
	// value is undefined. For the STRING type, the returned value.string is a
	// copy
	// of the input string. Remember to free() it after use!
	public static <V> V stringToValueConverter(final Class<V> type, final String valueString) {
		if (valueString == null) {
			return null; // NULL STRING.
		}

		if (type == Boolean.class) {
			// Boolean uses custom true/false strings.
			if (valueString.equals("true")) {
				return type.cast(Boolean.TRUE);
			}

			return type.cast(Boolean.FALSE);
		}
		else if (type == Byte.class) {
			return type.cast(Byte.parseByte(valueString));
		}
		else if (type == Short.class) {
			return type.cast(Short.parseShort(valueString));
		}
		else if (type == Integer.class) {
			return type.cast(Integer.parseInt(valueString));
		}
		else if (type == Long.class) {
			return type.cast(Long.parseLong(valueString));
		}
		else if (type == Float.class) {
			return type.cast(Float.parseFloat(valueString));
		}
		else if (type == Double.class) {
			return type.cast(Double.parseDouble(valueString));
		}
		else if (type == String.class) {
			return type.cast(valueString);
		}
		else {
			return null; // UNKNOWN TYPE.
		}
	}
}
